synapse/storage/databases/main/user_directory.py

```diff
@@ -259,7 +265,7 @@ def _get_next_batch(
                         if not self.hs.is_mine_id(user_id):
                             continue
 
-                        if self.get_if_app_services_interested_in_user(user_id):
+                        if not await self.should_include_local_user_in_dir(user_id):
                             continue
 
                         for other_user_id in users_with_profile:
```

Comment:
There's a hard-to-spot oversight right below here on -269.

We don't check to see if `other_user_id` ought to be excluded from the directory. We'll gleefully add them to `users_who_share_private_rooms`. This isn't a user-visible problem---if all of the logic elsewhere is correct, then `other_user_id` won't be in the `user_directory` table and so won't show up in searches. But it's a frustrating inconsistency.

Instead I'm going to filter out the excluded users early on, right before we assign a value to `users_with_profile`.